pitop/processing/algorithms/ball_detect.py

```diff
@@ -0,0 +1,247 @@
+from collections import deque
+import numpy as np
+from pitop.core import ImageFunctions
+from pitop.processing.utils.vision_functions import (
+    center_reposition,
+    get_control_angle,
+)
+from typing import Union
+from imutils import resize, grab_contours
+from imutils.video import FPS
+from pitop.core.data_stuctures import DotDict
+import atexit
+from pitop.processing.utils.vision_functions import import_opencv
+
+
+cv2 = import_opencv()
+
+colour_ranges = {
+    'red': [
+        {
+            'lower': (150, 100, 100),
+            'upper': (179, 255, 255)
+        },
+        {
+            'lower': (0, 100, 100),
+            'upper': (5, 255, 255)
+        }
+    ],
+    'green': [{
+        'lower': (60, 100, 100),
+        'upper': (90, 255, 255)
+    }],
+    'blue': [{
+        'lower': (100, 100, 100),
+        'upper': (130, 255, 255)
+    }]
+}
+
+draw_colour = {
+    'red': (0, 0, 255),
+    'green': (0, 255, 0),
+    'blue': (255, 0, 0)
+}
+
+ball_match_limits = {
+    'red': 0.1,  # red needs to be a higher limit since some skin types are redish in colour
+    'green': 0.05,
+    'blue': 0.05
+}
+
+MIN_BALL_RADIUS = 5
+BALL_CLOSE_RADIUS = 50
+BUFFER_LENGTH = 64
+detection_points = {
+    "red": deque(maxlen=BUFFER_LENGTH),
+    "green": deque(maxlen=BUFFER_LENGTH),
+    "blue": deque(maxlen=BUFFER_LENGTH)
+}
+
+
+class BallDetector:
+    def __init__(self,
+                 process_image_width: int = 320,
+                 input_format: str = "PIL",
+                 output_format: str = "PIL",
+                 print_fps: bool = False):
+        """
+        :param process_image_width: image width to scale to for image processing
+        :param input_format: input image format
+        :param output_format: output image format
+        """
+        self._process_image_width = process_image_width
+        self._input_format = input_format
+        self._output_format = output_format
+        self._frame_scaler = None
+        self._print_fps = print_fps
+        if self._print_fps:
+            self._fps = FPS().start()
+            atexit.register(self.__print_fps)
+
+    def detect(self, frame, colour: Union[str, tuple] = "red"):
```

Comment:
This function is way too long for me. It should be broken up

Reply:
See updates